inst/include/dust/cuda_launch_control.hpp

```diff
@@ -0,0 +1,233 @@
+#ifndef DUST_CUDA_LAUNCH_CONTROL_HPP
+#define DUST_CUDA_LAUNCH_CONTROL_HPP
+
+#include <dust/types.hpp>
+
+namespace dust {
+
+namespace cuda {
+
+struct launch_control {
+  size_t block_size;
+  size_t block_count;
+  size_t shared_size_bytes;
+  bool shared_int;
+  bool shared_real;
+};
+
+
+inline size_t device_shared_size(int device_id) {
+  int size = 0;
+#ifdef __NVCC__
+  if (device_id >= 0) {
+    CUDA_CALL(cudaDeviceGetAttribute(&size,
+                                     cudaDevAttrMaxSharedMemoryPerBlock,
+                                     device_id));
+  }
+#endif
+  return static_cast<size_t>(size);
+}
+
+
+// Tunable bits exposed to the front end
+class device_config {
+public:
+  device_config(int device_id, int run_block_size) :
+    device_id_(device_id),
+    run_block_size_(run_block_size),
+    shared_size_(device_shared_size(device_id_)),
+    enabled_(device_id_ >= 0) {
+#ifdef __NVCC__
+    if (device_id_ >= 0) {
+      CUDA_CALL(cudaSetDevice(device_id_));
+      CUDA_CALL(cudaDeviceSetCacheConfig(cudaFuncCachePreferL1));
+    }
+#endif
+  }
+
+  const int device_id_;
```

Comment:
```suggestion
private:
  const int device_id_;
```

Reply:
or this is a struct not a class?

Reply:
we refer to these later - it's basically a struct with a nontrivial constructor (the only difference between a struct and a class is the default visibility of the members, really)